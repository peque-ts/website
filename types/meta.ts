export interface ProjectMeta {
  title: string;
  description: string;
}

export interface SectionMeta {
  title: string;
  description: string;
  order: number;
}
