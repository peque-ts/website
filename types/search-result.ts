export interface SearchResult {
  title: string;
  description: string;
  link: string;
}
