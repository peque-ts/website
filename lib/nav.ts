import matter from 'gray-matter';
import fs from 'node:fs/promises';
import path from 'node:path';

import { read } from './fs';
import { NavItem, PrevNextNavItems } from './nav.types';

async function buildSideNavItems(project: string, activeSection: string): Promise<NavItem[]> {
  const items: NavItem[] = [];

  const sectionFiles = await fs.readdir(path.join(process.cwd(), 'docs', project));

  for (const sectionFile of sectionFiles) {
    const markdown = await read(`docs/${project}/${sectionFile}`);

    const sectionSlug = sectionFile.replace(/.md$/, '');
    const { title, order } = matter(markdown).data;

    items.push({
      name: title,
      to: `/docs/${project}/${sectionSlug}`,
      order,
      active: sectionSlug === activeSection,
      slug: sectionSlug,
    });
  }

  return items.sort((a, b) => {
    return a.order > b.order ? 1 : a.order < b.order ? -1 : 0;
  });
}

function getPrevNextNavItems(sideNavItems: NavItem[], activeSection: string): PrevNextNavItems {
  const currentItem = sideNavItems.find((item) => item.slug === activeSection);

  if (!currentItem) {
    throw new Error(
      `Cannot find navigation item with slug = [${activeSection}]. This looks like a bug.`,
    );
  }

  return {
    prev: sideNavItems.find((item) => item.order === currentItem.order - 1) ?? null,
    next: sideNavItems.find((item) => item.order === currentItem.order + 1) ?? null,
  };
}

export { buildSideNavItems, getPrevNextNavItems };
