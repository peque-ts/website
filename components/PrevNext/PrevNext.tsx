import clsx from 'clsx';
import Link from 'next/link';
import React from 'react';

import { PrevNextNavItems } from '../../lib/nav.types';
import { Icon } from '../Icon';

interface Props {
  items: PrevNextNavItems;
}

export const PrevNext: React.VFC<Props> = ({ items }) => {
  const renderButton = (type: 'prev' | 'next') => {
    const item = items[type];

    return (
      <div className={clsx('w-full desktop:w-80', { 'hidden laptop:block': type === 'prev' })}>
        {!!item && (
          <Link href={item.to}>
            <a
              className={clsx(
                'flex items-center font-semibold justify-between rounded p-4 transition bg-secondary-700 hover:bg-secondary-800',
                {
                  'flex-row text-right': type === 'prev',
                  'flex-row-reverse': type === 'next',
                },
              )}
            >
              <div className={clsx('flex', type === 'next' ? 'rotate-180 ml-4' : 'mr-4')}>
                <Icon alt={`Go to ${item.name}`} name="arrow-left" size={24} />
              </div>
              <div
                className={clsx(
                  'flex flex-col space-y-1',
                  type === 'next' ? 'items-start' : 'items-end',
                )}
              >
                <span className="text-sm text-secondary-200">
                  {type === 'prev' ? 'Prev' : 'Next'}
                </span>
                <span className="text-base text-white">{item.name}</span>
              </div>
            </a>
          </Link>
        )}
      </div>
    );
  };

  return (
    <div className="flex mt-4 laptop:justify-between laptop:space-x-4 desktop:space-x-0">
      {renderButton('prev')}
      {renderButton('next')}
    </div>
  );
};
