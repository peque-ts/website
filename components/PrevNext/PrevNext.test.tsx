import { render } from '@testing-library/react';

import { PrevNextNavItems } from '../../lib/nav.types';
import { PrevNext } from './PrevNext';

describe('PrevNext', () => {
  const testItems: PrevNextNavItems[] = [
    {
      prev: null,
      next: {
        to: '/last',
        name: 'Last',
        slug: 'last',
        order: 2,
        active: false,
      },
    },
    {
      prev: {
        to: '/first',
        name: 'First',
        slug: 'first',
        order: 1,
        active: false,
      },
      next: {
        to: '/last',
        name: 'Last',
        slug: 'last',
        order: 3,
        active: false,
      },
    },
    {
      prev: {
        to: '/first',
        name: 'First',
        slug: 'first',
        order: 1,
        active: false,
      },
      next: null,
    },
  ];

  it.each(testItems)('should render', (items) => {
    const { asFragment } = render(<PrevNext items={items} />);

    expect(asFragment()).toMatchSnapshot();
  });
});
