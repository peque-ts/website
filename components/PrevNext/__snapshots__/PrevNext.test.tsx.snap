// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrevNext should render 1`] = `
<DocumentFragment>
  <div
    class="mt-4 flex laptop:justify-between laptop:space-x-4 desktop:space-x-0"
  >
    <div
      class="w-full desktop:w-80 hidden laptop:block"
    />
    <div
      class="w-full desktop:w-80"
    >
      <a
        class="flex items-center justify-between rounded bg-secondary-700 p-4 font-semibold transition hover:bg-secondary-800 flex-row-reverse"
        href="/last"
      >
        <div
          class="flex text-secondary-200 ml-4 rotate-180"
        >
          <svg
            height="24"
            viewBox="0 0 64 64"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10.15.5 4 6.65a.48.48 0 0 0 0 .7l6.15 6.15"
              fill="none"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              transform="rotate(.045) scale(4.57143)"
            />
          </svg>
        </div>
        <div
          class="flex flex-col space-y-1 items-start"
        >
          <span
            class="text-sm text-secondary-200"
          >
            Next
          </span>
          <span
            class="text-base text-white"
          >
            Last
          </span>
        </div>
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`PrevNext should render 2`] = `
<DocumentFragment>
  <div
    class="mt-4 flex laptop:justify-between laptop:space-x-4 desktop:space-x-0"
  >
    <div
      class="w-full desktop:w-80 hidden laptop:block"
    >
      <a
        class="flex items-center justify-between rounded bg-secondary-700 p-4 font-semibold transition hover:bg-secondary-800 flex-row text-right"
        href="/first"
      >
        <div
          class="flex text-secondary-200 mr-4"
        >
          <svg
            height="24"
            viewBox="0 0 64 64"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10.15.5 4 6.65a.48.48 0 0 0 0 .7l6.15 6.15"
              fill="none"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              transform="rotate(.045) scale(4.57143)"
            />
          </svg>
        </div>
        <div
          class="flex flex-col space-y-1 items-end"
        >
          <span
            class="text-sm text-secondary-200"
          >
            Prev
          </span>
          <span
            class="text-base text-white"
          >
            First
          </span>
        </div>
      </a>
    </div>
    <div
      class="w-full desktop:w-80"
    >
      <a
        class="flex items-center justify-between rounded bg-secondary-700 p-4 font-semibold transition hover:bg-secondary-800 flex-row-reverse"
        href="/last"
      >
        <div
          class="flex text-secondary-200 ml-4 rotate-180"
        >
          <svg
            height="24"
            viewBox="0 0 64 64"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10.15.5 4 6.65a.48.48 0 0 0 0 .7l6.15 6.15"
              fill="none"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              transform="rotate(.045) scale(4.57143)"
            />
          </svg>
        </div>
        <div
          class="flex flex-col space-y-1 items-start"
        >
          <span
            class="text-sm text-secondary-200"
          >
            Next
          </span>
          <span
            class="text-base text-white"
          >
            Last
          </span>
        </div>
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`PrevNext should render 3`] = `
<DocumentFragment>
  <div
    class="mt-4 flex laptop:justify-between laptop:space-x-4 desktop:space-x-0"
  >
    <div
      class="w-full desktop:w-80 hidden laptop:block"
    >
      <a
        class="flex items-center justify-between rounded bg-secondary-700 p-4 font-semibold transition hover:bg-secondary-800 flex-row text-right"
        href="/first"
      >
        <div
          class="flex text-secondary-200 mr-4"
        >
          <svg
            height="24"
            viewBox="0 0 64 64"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10.15.5 4 6.65a.48.48 0 0 0 0 .7l6.15 6.15"
              fill="none"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              transform="rotate(.045) scale(4.57143)"
            />
          </svg>
        </div>
        <div
          class="flex flex-col space-y-1 items-end"
        >
          <span
            class="text-sm text-secondary-200"
          >
            Prev
          </span>
          <span
            class="text-base text-white"
          >
            First
          </span>
        </div>
      </a>
    </div>
    <div
      class="w-full desktop:w-80"
    />
  </div>
</DocumentFragment>
`;
